File: Java/Medium/271-EncodeAndDecodeString.java
// Description: https://leetcode.com/problems/encode-and-decode-strings/description/

/*
    Approach: Length Prefix Encoding

    Encode: Prefix each string with its length followed by '#' delimiter.
    Also prefix the entire message with the list size and '#'.
    This handles any characters within strings including special characters.
    Decode: First extract list size, then iteratively extract each string.
    For each string, read the length before '#', then extract that many characters.
    This encoding is unambiguous and handles empty strings correctly.

    Time complexity: O(n) - Where n is total characters across all strings.
    Space complexity: O(n) - For the encoded/decoded strings.
*/

import java.util.*;
class Solution {

    public String encode(List<String> strs) {
        StringBuilder encodedString = new StringBuilder();
        encodedString.append(strs.size()).append("#");

        for (String str : strs) {
            encodedString.append(str.length()).append("#").append(str);
        }

        return encodedString.toString();
    }

    public List<String> decode(String s) {
        int currentPos = 0;
        int hashPos = s.indexOf('#', currentPos);
        int listSize = Integer.parseInt(s.substring(currentPos, hashPos));
        currentPos = hashPos + 1;

        List<String> decodedStrings = new ArrayList<>(listSize);

        for (int i = 0; i < listSize; i++) {
            hashPos = s.indexOf('#', currentPos);
            int wordSize = Integer.parseInt(s.substring(currentPos, hashPos));
            currentPos = hashPos + 1;
            decodedStrings.add(s.substring(currentPos, currentPos + wordSize));
            currentPos += wordSize;
        }

        return decodedStrings;
    }
}

// Test
class Test {
    public static String formatList(List<String> list) {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < list.size(); i++) {
            sb.append("\"").append(list.get(i)).append("\"");
            if (i < list.size() - 1) sb.append(", ");
        }
        sb.append("]");
        return sb.toString();
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // Example 1: ["hello", "world"]
        List<String> strs1 = Arrays.asList("hello", "world");
        String encoded1 = solution.encode(strs1);
        List<String> decoded1 = solution.decode(encoded1);
        System.out.println("Output: " + formatList(decoded1));

        // Example 2: ["", "a", "abc", "defg"]
        List<String> strs2 = Arrays.asList("", "a", "abc", "defg");
        String encoded2 = solution.encode(strs2);
        List<String> decoded2 = solution.decode(encoded2);
        System.out.println("Output: " + formatList(decoded2));
    }
}

File: C++/Medium/271-EncodeAndDecodeString.cc
// Description: https://leetcode.com/problems/encode-and-decode-strings/description/

/*
    Approach: Length Prefix Encoding

    Encode: Prefix each string with its length followed by '#' delimiter.
    Also prefix the entire message with the list size and '#'.
    This handles any characters within strings including special characters.
    Decode: First extract list size, then iteratively extract each string.
    For each string, read the length before '#', then extract that many characters.
    This encoding is unambiguous and handles empty strings correctly.

    Time complexity: O(n) - Where n is total characters across all strings.
    Space complexity: O(n) - For the encoded/decoded strings.
*/

#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    string encode(vector<string>& strs) {
        string encodedString = to_string(strs.size()) + "#";
        for (const string& str : strs) encodedString += to_string(str.size()) + '#' + str;
        return encodedString;
    }

    vector<string> decode(string s) {
        size_t currentPos = 0;
        size_t hashPos = s.find('#', currentPos);
        unsigned int listSize = stoi(s.substr(currentPos, hashPos - currentPos));
        currentPos = hashPos + 1;

        vector<string> decodedStrings(listSize);

        for (unsigned i = 0; i < listSize; ++i) {
            hashPos = s.find('#', currentPos);
            int wordSize = stoi(s.substr(currentPos, hashPos - currentPos));
            currentPos = hashPos + 1;
            decodedStrings[i] = s.substr(currentPos, wordSize);
            currentPos += wordSize;
        }
        return decodedStrings;
    }
};

// Test
void printDecoded(const vector<string>& decoded) {
    cout << "[";
    for (size_t i = 0; i < decoded.size(); i++) {
        cout << "\"" << decoded[i] << "\"";
        if (i < decoded.size() - 1) cout << ", ";
    }
    cout << "]";
}

int main() {
    Solution solution;

    // Example 1
    vector<string> strs1 = {"hello", "world"};
    cout << "Input: [\"hello\", \"world\"]" << endl;
    string encoded1 = solution.encode(strs1);
    vector<string> decoded1 = solution.decode(encoded1);
    cout << "Output: ";
    printDecoded(decoded1);
    cout << endl << endl;

    // Example 2
    vector<string> strs2 = {"", "a", "abc", "defg"};
    cout << "Input: [\"\", \"a\", \"abc\", \"defg\"]" << endl;
    string encoded2 = solution.encode(strs2);
    vector<string> decoded2 = solution.decode(encoded2);
    cout << "Output: ";
    printDecoded(decoded2);
    cout << endl << endl;

    return 0;
}
